### DISCUSSION TITLE
How to add Quarkus features required for a KC extension ?

### DISCUSSION THREAD

**Discussion by hrenard** [keycloak/keycloak#10729] [Archive]
Hi,

I'm trying to port a Keycloak extension to the new Quarkus runtime. The extension successfully load but fail to work because the `org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder` is not found.
If I understand correctly, I need to add `rest-client` add `rest-client-jackson` to the runtime. But I fail to understand how.
Do I  need to add extra config while building ?
Do I need to migrate the code to a Quarkus extension format ?
Is there an easy way to do this ?

**Comment by sventorben** [comment]
Hey @hrenard ,
I am facing the same issue and already tried to migrate the code to the microprofile api (https://github.com/eclipse/microprofile-rest-client), but without success. Did you make any progress with this?

**Reply by hrenard** [reply]
No, I saw [this](https://github.com/keycloak/keycloak/pull/10714) and was hoping my problem would be fixed in the next release.

**Comment by errier** [comment]
I'm in the process of migrating from Wildfly to Quarkus, and I drop any dependency I need for a provider in the providers directory and then run build.
Edit: Actually, using a third party library is documented here: https://www.keycloak.org/server/configuration-provider#_using_third_party_dependencies

**Comment by hongquyen196** [comment]
Hi @hrenard, I also have the same problem with you when changing to use Keycloak X (17.0.0).

But I follow document https://www.keycloak.org/docs/latest/server_development/#registering-provider-implementations
Keycloak writes "**If your provider needs additional dependencies not already provided by Keycloak copy these to the providers directory**."

I have downloaded 2 files:
https://repo1.maven.org/maven2/org/jboss/resteasy/resteasy-client/4.7.5.Final/resteasy-client-4.7.5.Final.jar https://repo1.maven.org/maven2/org/jboss/resteasy/resteasy-client-api/4.7.5.Final/resteasy-client-api-4.7.5.Final.jar
Then I copy all these third-party libraries to /opt/keycloak/providers and it works. :)

![image](https://user-images.githubusercontent.com/25485764/162142789-f5588a57-ac89-4df2-8a3d-a099e140f271.png)

Good luck!

**Reply by hrenard** [reply]
Hi, I'm using this to bundle all in one jar :
```xml
<plugin>
  <artifactId>maven-assembly-plugin</artifactId>
  <version>2.6</version>
  <executions>
    <execution>
      <id>make-assembly</id>
      <phase>package</phase>
      <goals>
        <goal>single</goal>
      </goals>
      <configuration>
        <archive>
          <manifest>
            <mainClass>${main.class}</mainClass>
          </manifest>
        </archive>
        <descriptorRefs>
          <descriptorRef>jar-with-dependencies</descriptorRef>
        </descriptorRefs>
      </configuration>
    </execution>
  </executions>
</plugin>
```
But when I tried to run a bundle with resteasy it crashed, apparently because of a conflict. Maybe I included to many or to few dependencies. I will try this also.

**Reply by hongquyen196** [reply]
Oh, I am using this plugin too.
But you should use "provided" instead of "compile" for all third-party dependencies.

```
<dependency>
	<groupId>org.keycloak</groupId>
	<artifactId>keycloak-admin-client</artifactId>
	<!--Keycloak 17.0.0-legacy doesn't work with scope "provided"-->
	<scope>compile</scope>
</dependency>
<dependency>
	<groupId>org.jboss.resteasy</groupId>
	<artifactId>resteasy-client</artifactId>
	<scope>provided</scope>
	<version>${resteasy.version}</version>
</dependency>
```
This is my pom file for my provider. It works with both Keycloak 17.0.0-legacy and Keycloak 17.0.0 (X).

Note: It's will not work on the new Keycloak 17.0.1 (I am testing it).

**Reply by hrenard** [reply]
Thanks ! I know about provided scope. I probably just need to spend some time to verify my dependencies and dependencies of dependencies... But if it works for your, it'll be worth it.

**Reply by maximepiton** [reply]
Hi @hongquyen196,
I also have issues while trying to make SPIs using keycloak-admin-client work with quarkus.
I'm not sure I understand, do you both manually put the resteasy-client/resteasy-client-api/keycloak-admin-client JARs in `/etc/keycloak/providers`, AND use `maven-assembly-plugin`?
By any chance, could you share your working `pom.xml`?
Cheers

**Reply by hongquyen196** [reply]
> Hi @hongquyen196, I also have issues while trying to make SPIs using keycloak-admin-client work with quarkus. I'm not sure I understand, do you both manually put the resteasy-client/resteasy-client-api/keycloak-admin-client JARs in `/etc/keycloak/providers`, AND use `maven-assembly-plugin`? By any chance, could you share your working `pom.xml`? Cheers
Yes, I copied 3 files resteasy-client, resteasy-client-api, keycloak-admin-client JARs into **/opt/keycloak/providers/**. Also used   `maven-assembly-plugin`.
PLEASE MAKE SURE THAT ONLY USE RESTEASY 4.7.5 Final
You can see pom content above.


> ```
> <dependency>
> 	<groupId>org.keycloak</groupId>
> 	<artifactId>keycloak-admin-client</artifactId>
> 	
> 	<scope>compile</scope>
> </dependency>
> <dependency>
> 	<groupId>org.jboss.resteasy</groupId>
> 	<artifactId>resteasy-client</artifactId>
> 	<scope>provided</scope>
> 	<version>${resteasy.version}</version>
> </dependency>
> ```



> ```
> <plugin>
>   <artifactId>maven-assembly-plugin</artifactId>
>   <version>2.6</version>
>   <executions>
>     <execution>
>       <id>make-assembly</id>
>       <phase>package</phase>
>       <goals>
>         <goal>single</goal>
>       </goals>
>       <configuration>
>         <archive>
>           <manifest>
>             <mainClass>${main.class}</mainClass>
>           </manifest>
>         </archive>
>         <descriptorRefs>
>           <descriptorRef>jar-with-dependencies</descriptorRef>
>         </descriptorRefs>
>       </configuration>
>     </execution>
>   </executions>
> </plugin>
> ```



Good luck!

**Reply by maximepiton** [reply]
Thanks @hongquyen196, it worked!

**Comment by zampettim** [comment]
I'm running into similar issues. One thing is that I'm running with Java 11. What version of Java was the working examples using?

**Reply by hongquyen196** [reply]
I am using Java 11.

**Comment by hrenard** [comment]
Hey, I still failed to make this work. Can some here help me out :sweat_smile: ?  The [extension](https://lab.libreho.st/libre.sh/scim/keycloak-scim) I try to port use [this package](https://github.com/pingidentity/scim2) which needs a `javax.ws.rs.client.Client`.

**Comment by hrenard** [comment]
I just found the solution [here](https://github.com/keycloak/keycloak/issues/11973#issuecomment-1127824518) ! `org.wildfly.client.wildfly-client-config` is required. 
Thanks, @hongquyen196 !
